src/policy/ancestor_packages.h

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) 2021-2022 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_POLICY_ANCESTOR_PACKAGES_H
+#define BITCOIN_POLICY_ANCESTOR_PACKAGES_H
+
+#include <policy/feerate.h>
+#include <policy/packages.h>
+
+#include <vector>
+
+/** A potential BIP331 Ancestor Package, i.e. one transaction with its set of ancestors.
+ * This class does not have any knowledge of chainstate, so it cannot determine whether all
+ * unconfirmed ancestors are present. Its constructor accepts any list of transactions that
+ * IsConsistent(), linearizes them topologically, and determines whether it IsAncestorPackage().
+ * If fee and vsizes are given for each transaction, it can also linearize the transactions using
+ * the ancestor score-based mining algorithm via MiniMiner.
+ *
+ * Skip() and SkipWithDescendants() can be used to omit transactions. Txns() returns all
+ * transactions linearized, and FilteredTxns() is the same but excludes the skipped ones.
+ * GetAncestorSet() can be used to get a transaction's "subpackage," i.e. ancestor set within the
+ * package, also excluding skipped ones.
+ * */
+class AncestorPackage
+{
+    /** Whether m_txns contains a connected package in which all transactions are ancestors of the
+     * last transaction. This object is not aware of chainstate. So if m_txns only includes a
+     * grandparent and not the "connecting" parent, this will (incorrectly) determine that the
+     * grandparent is not an ancestor.
+     * */
+    bool m_is_ancestor_package{false};
+
+    /** Linearized transactions. Topological (IsSorted()) or, if fee information is provided through
+     * LinearizeWithFees(), using ancestor set scores. */
+    Package m_txns;
+
+    struct PackageEntry {
+        /** Whether this transaction should be skipped in GetAncestorSet() and linearization by
+         * fees, i.e. because it is already in the mempool.
+         * This value can be set to true by calling Skip(). */
+        bool skip{false};
+        /** Whether this transaction "dangles," i.e. we know nothing about it it because it is
+         * missing inputs or depends on another transaction that is missing inputs.
+         * This value can be set to true by calling SkipWithDescendants(). */
+        bool dangles{false};
+        /** This value starts as std::nullopt when we don't have any fee information yet. It can be
+         * updated by calling LinearizeWithFees() if this entry isn't being skipped. */
+        std::optional<uint32_t> mining_sequence;
+        /** (Modified) fees of this transaction. Starts as std::nullopt, can be updated using AddFeeAndVsize(). */
+        std::optional<CAmount> fee;
```

Comment:
Unclear if “(Modified) fees” means `GetModifiedFee()`

Reply:
Removed mention of "modified," as it's not really relevant whether it's real fees or not